File: src/main/java/com/lykke/matching/engine/daos/MarketOrder.java
package com.lykke.matching.engine.daos;

import java.util.Date;

public class MarketOrder extends Order {
    //partition key: client_id
    //row key: uid

    //date of execution
    Date matchedAt;
    boolean straight;

    Double dustSize;

    public MarketOrder() {
    }

    public MarketOrder(String uid, String assetPairId, String clientId, Double volume, Double price, String status,
                       Date createdAt, Date registered, String transactionId, Date matchedAt, boolean straight) {
        this(uid, assetPairId, clientId, volume, price, status, createdAt, registered, transactionId, matchedAt, straight, 0.0d);
    }

    public MarketOrder(String uid, String assetPairId, String clientId, Double volume, Double price, String status,
                       Date createdAt, Date registered, String transactionId, Date matchedAt, boolean straight, Double dustSize) {
        super(uid, assetPairId, clientId, volume, price, status, createdAt, registered, transactionId);
        this.matchedAt = matchedAt;
        this.straight = straight;
        this.dustSize = dustSize;
    }

    public boolean isBuySide() {
        return straight ? super.isBuySide() : !super.isBuySide();
    }

    public boolean isOrigBuySide() {
        return super.isBuySide();
    }

    @Override
    public String toString() {
        return "MarketOrder{"  +
                "assetPairId='" + assetPairId + '\'' +
                ", clientId='" + clientId + '\'' +
                ", volume=" + volume +
                ", dustSize=" + dustSize +
                ", price=" + price +
                ", straight=" + straight +
                ", status='" + status + '\'' +
                ", createdAt=" + createdAt +
                ", registered=" + registered +
                ", transactionId='" + transactionId + '\'' +
                ", matchedAt=" + matchedAt +
                '}';
    }

    public Date getMatchedAt() {
        return matchedAt;
    }

    public void setMatchedAt(Date matchedAt) {
        this.matchedAt = matchedAt;
    }

    public boolean getStraight() {
        return straight;
    }

    public void setStraight(boolean straight) {
        this.straight = straight;
    }

    public Double getDustSize() {
        return dustSize;
    }

    public void setDustSize(Double dustSize) {
        this.dustSize = dustSize;
    }
}

File: src/main/kotlin/com/lykke/matching/engine/daos/MarketOrder.kt
package com.lykke.matching.engine.daos

import com.lykke.matching.engine.daos.fee.v2.NewFeeInstruction
import java.math.BigDecimal
import com.lykke.matching.engine.utils.NumberUtils
import com.lykke.matching.engine.daos.v2.FeeInstruction
import java.util.Date

class MarketOrder(id: String,
                  uid: String,
                  assetPairId: String,
                  clientId: String,
                  volume: BigDecimal,
                  var price: BigDecimal?,
                  status: String,
                  statusDate: Date?,
                  createdAt: Date,
                  registered: Date?,
                  var matchedAt: Date?,
                  val straight: Boolean,
                  reservedLimitVolume: BigDecimal? = null,
                  fee: FeeInstruction? = null,
                  fees: List<NewFeeInstruction>? = null)
    : Order(id, uid, assetPairId, clientId, volume, status, createdAt, registered, reservedLimitVolume, fee, fees, statusDate) {

    override fun isOrigBuySide(): Boolean {
        return super.isBuySide()
    }

    override fun isBuySide(): Boolean {
        return if (straight) super.isBuySide() else !super.isBuySide()
    }

    override fun isStraight(): Boolean {
        return straight
    }

    override fun calculateReservedVolume(): BigDecimal {
        return reservedLimitVolume ?: BigDecimal.ZERO
    }

    override fun updateMatchTime(time: Date) {
        matchedAt = time
    }

    override fun takePrice(): BigDecimal? {
        return price
    }

    override fun updatePrice(price: BigDecimal) {
        this.price = price
    }

    override fun updateRemainingVolume(volume: BigDecimal) {
        //nothing to do
    }

    override fun copy(): MarketOrder {
        return MarketOrder(id, externalId, assetPairId, clientId, volume, price, status, statusDate!!, createdAt, registered, matchedAt, straight, reservedLimitVolume, fee, fees)
    }

    override fun applyToOrigin(origin: Copyable) {
        super.applyToOrigin(origin)
        origin as MarketOrder
        origin.price = price
        origin.matchedAt = matchedAt
    }

    override fun toString(): String {
        return "market order id: $id, client: $clientId, asset: $assetPairId, " +
                "volume: ${NumberUtils.roundForPrint(volume)}, straight: $straight"
    }
}